src/core/namespace.c

```diff
@@ -2245,32 +2252,69 @@ int setup_namespace(const NamespaceParameters *p, char **error_path) {
         if (r < 0)
                 return r;
 
-        if (p->tmp_dir) {
-                bool ro = streq(p->tmp_dir, RUN_SYSTEMD_EMPTY);
+        if (p->private_tmp_as_tmpfs) {
+                _cleanup_free_ char *tmp_dir = NULL, *var_tmp_dir = NULL;
+                MountEntry *tmp_entry, *var_tmp_entry;
 
-                MountEntry *me = mount_list_extend(&ml);
-                if (!me)
+                tmp_dir = path_join(p->private_tmp_dir, "tmp");
+                var_tmp_dir = path_join(p->private_tmp_dir, "var/tmp");
+                if (!tmp_dir || !var_tmp_dir)
                         return log_oom_debug();
 
-                *me = (MountEntry) {
+                tmp_entry = mount_list_extend(&ml);
+                var_tmp_entry = mount_list_extend(&ml);
+                if (!tmp_entry || !var_tmp_entry)
+                        return log_oom_debug();
```

Comment:
hmm, it feels weird allocating this, but not initializing it right away... it think this should be swapped around

Reply:
i still think aht extending the list, and initializing the new item would be much preferable, because otherwise you end up freeing uninitialized items if the last extension fails (i mean, it's not actually uninitialized, just initialized to defaults, but still...)